### PR TITLE
feat: report scan progress in Snyk Code [ROAD-1259]

### PR DESCRIPTION
**Pull request by asaf92**
### Description

Send progress updates via `$/Snyk.scan` messages in Snyk Code

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] README.md updated, if user-facing

### CODE REVIEW

File: application/server/lsp/message_types.go [hunk 1]
```diff
@@ -680,3 +680,22 @@ const Log MessageType = 4
 type SnykTrustedFoldersParams struct {
 	TrustedFolders []string `json:"trustedFolders"`
 }
+
+type ScanStatus string
+
+const (
+	Initial     ScanStatus = "initial"
```
Comment (asaf92) [inline application/server/lsp/message_types.go:687]: Remove "initial"

File: README.md [hunk 2]
```diff
@@ -79,12 +79,25 @@ Right now the language server supports the following actions:
...
     "trustedFolders": ["/a/path/to/trust"]
   }
   ```
 
+- Scan Notification
+  - method: `$/snyk.scan`
+  - payload:
+  ```json5
+  {
+    "status": "InProgress", // possible values: "Error", "In Progress", "Success"
```
Comment (michelkaporin) [inline README.md:93]: Let's update them

File: internal/notification/notifier_mock.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+			Severity:        lsp.DiagnosticsSeverityWarning,
+			Code:            "Snyk Error",
+			CodeDescription: lsp.CodeDescription{Href: "https://snyk.io/user-hub"},
+			Message:         err.Error(),
+		}},
+	}
+	m.sentMessages = append(m.sentMessages, msg)
+}
+
+func (m *MockNotifier) GetSendShowMessageCounter() int { return m.sendShowMessageCounter }
```
Comment (michelkaporin) [inline internal/notification/notifier_mock.go:60]: `func (m *MockNotifier) ShowMessageCounter()`

File: internal/notification/scannotifier.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	SendError(folderPath string)
+}
+
+type scanNotifier struct {
+	notifier    Notifier
+	productName string
+}
+
+func NewScanNotifier(notifier Notifier, productName string) ScanNotifier {
+	return &scanNotifier{
```
Comment (michelkaporin) [inline internal/notification/scannotifier.go:17]: Let's add some validation to the type

File: internal/notification/scannotifier_mock.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+func (m *MockScanNotifier) SendSuccess(folderPath string) {
+	m.successCalls = append(m.successCalls, folderPath)
+}
+
+func (m *MockScanNotifier) SendError(folderPath string) {
+	m.errorCalls = append(m.errorCalls, folderPath)
+}
+
+func (m *MockScanNotifier) GetInProgressCalls() []string {
```
Comment (michelkaporin) [inline internal/notification/scannotifier_mock.go:25]: Code style

File: internal/notification/scannotifier_test.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
+package notification_test
```
Comment (michelkaporin) [inline internal/notification/scannotifier_test.go:1]: file name

File: internal/notification/scannotifier_test.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+	lsp2 "github.com/snyk/snyk-ls/application/server/lsp"
+	"github.com/snyk/snyk-ls/internal/notification"
+	"github.com/snyk/snyk-ls/internal/testutil"
+)
+
+const productName = "foo"
+const folderPath = "/test/folderPath"
+
+func Test_SendInProgressMessage_InProgressMessageSent(t *testing.T) {
```
Comment (michelkaporin) [inline internal/notification/scannotifier_test.go:16]: Maybe use a matrix for assertion?

File: internal/notification/scannotifier.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
+package notification
+
+import "github.com/snyk/snyk-ls/application/server/lsp"
+
+type ScanNotifier interface {
```
Comment (michelkaporin) [inline internal/notification/scannotifier.go:5]: Maybe move this to snyk domain?

File: infrastructure/code/bundle.go [hunk 9]
```diff
@@ -110,15 +116,18 @@ func (b *Bundle) retrieveAnalysis(ctx context.Context) []snyk.Issue {
...
 
 		if status.message == "COMPLETE" {
-			log.Trace().Str("method", "retrieveAnalysis").Str("requestId", b.requestId).
+			logger.Trace().Str("requestId", b.requestId).
 				Msg("sending diagnostics...")
 			p.End("Analysis complete.")
 			return issues
+		} else if status.message == "ANALYZING" {
+			logger.Trace().Msg("\"Analyzing\" message received, sending In-Progress message to client")
+			//b.scanNotifier.SendInProgress(b.rootPath)
```
Comment (michelkaporin) [inline infrastructure/code/bundle.go:125]: Let's remove redundant comment

File: infrastructure/code/code_test.go [hunk 10]
```diff
@@ -307,9 +367,65 @@ func Test_CodeScanRunning_ScanCalled_ScansRunSequentially(t *testing.T) {
...
+	_, mockScanNotifier, scanner := setupTestScanner()
+
+	// Act
+	scanner.Scan(context.Background(), "", tempDir)
+
+	// Assert
+	assert.NotEmpty(t, mockScanNotifier.GetInProgressCalls())
+}
+
+//func Test_AnalyzingMessageReturned_InProgressMessageSentToClient(t *testing.T) {
```
Comment (michelkaporin) [inline infrastructure/code/code_test.go:383]: Do we need it?

File: infrastructure/code/code_test.go [hunk 11]
```diff
@@ -61,176 +62,237 @@ func setupDocs() (string, lsp.TextDocumentItem, lsp.TextDocumentItem, []byte, []
...
 		&snyk_api.FakeApiClient{CodeEnabled: true},
 		error_reporting.NewTestErrorReporter(),
 		ux2.NewTestAnalytics(),
+		notification.NewScanNotifier(notification.NewNotifier(), "code"),
 	)
 
 	return snykCodeMock, scanner
 }
 
+func setupTestScanner() (*FakeSnykCodeClient, *notification.MockScanNotifier, *Scanner) {
```
Comment (michelkaporin) [inline infrastructure/code/code_test.go:164]: Why having two functions, when you can parametrise the above one to use mocknotifier?

File: infrastructure/code/fake_snyk_code_api_service.go [hunk 12]
```diff
@@ -212,7 +229,16 @@ func (f *FakeSnykCodeClient) RunAnalysis(
 	FakeSnykCodeApiServiceMutex.Unlock()
 
 	issues := []snyk.Issue{FakeIssue}
+	if f.AnalyzingMessageCount > 0 {
```
Comment (michelkaporin) [inline infrastructure/code/fake_snyk_code_api_service.go:232]: Is this needed?

File: infrastructure/code/code_test.go (hunk 11 above)
Comment (asaf92) [reply infrastructure/code/code_test.go:164]: The original `setupTestScanner` is used for tests that assert results by actively listening to the messages in `notification`, so I'd have to refactor many tests to do it. Adding a parameter also requires to refactor all the calls to the setup function.

**Pull request merged**